rotkehlchen/exchanges/bitstamp.py

```diff
@@ -0,0 +1,711 @@
+    https://github.com/kmadac/bitstamp-python-client
+    """
+    def __init__(
+            self,
+            api_key: ApiKey,
+            secret: ApiSecret,
+            database: 'DBHandler',
+            msg_aggregator: MessagesAggregator,
+    ):
+        super(Bitstamp, self).__init__('bitstamp', api_key, secret, database)
+        self.base_uri = 'https://www.bitstamp.net/api'
+        self.msg_aggregator = msg_aggregator
+        # NB: X-Auth-Signature, X-Auth-Nonce, X-Auth-Timestamp and Content-Type per request
+        self.session.headers.update({
+            'X-Auth': f'BITSTAMP {self.api_key}',
+            'X-Auth-Version': 'v2',
+        })
+
+    @protect_with_lock()
+    @cache_response_timewise()
+    def query_balances(self) -> Tuple[Optional[Dict[Asset, Balance]], str]:
+        """Return the account balances on Bistamp
+
+        The balance endpoint returns a dict where the keys (str) are related to
+        assets and the values (str) amounts. The keys that end with `_balance`
+        contain the exact amount of an asset the account is holding (available
+        amount + orders amount, per asset).
+        """
+        response = self._api_query('balance')
+
+        if response.status_code != HTTPStatus.OK:
+            result, msg = self._process_unsuccessful_response(
+                response=response,
+                case='balances',
+            )
+            return result, msg
+        try:
+            response_dict = rlk_jsonloads_dict(response.text)
+        except JSONDecodeError:
+            msg = f'Bitstamp returned invalid JSON response: {response.text}.'
+            log.error(msg)
+            raise RemoteError(msg)
+
+        asset_balance: Dict[Asset, Balance] = {}
+        for entry, amount in response_dict.items():
+            amount = FVal(amount)
+            if not entry.endswith('_balance') or amount == ZERO:
+                continue
+
+            symbol = entry.split('_')[0]  # If no `_`, defaults to entry
+            try:
+                asset = Asset(symbol)
+            except UnknownAsset as e:
+                log.error(str(e))
+                self.msg_aggregator.add_warning(
+                    f'Found unknown Bistamp asset {e.asset_name}. Ignoring its balance query.',
+                )
+                continue
+            try:
+                usd_price = Inquirer().find_usd_price(asset=asset)
+            except RemoteError as e:
+                log.error(str(e))
+                self.msg_aggregator.add_error(
+                    f'Error processing Bitstamp balance result due to inability to '
+                    f'query USD price: {str(e)}. Skipping balance entry.',
+                )
+                continue
+
+            asset_balance[asset] = Balance(
+                amount=amount,
+                usd_value=amount * usd_price,
+            )
+
+        return asset_balance, ''
+
+    def query_online_deposits_withdrawals(
```

Comment:
@LefterisJP at the moment it only requests the `user_transactions` endpoint, which is not capable of returning the asset movement `address`, nor `transaction_id`. These would require to request as well `crypto_transactions` and do a mix. At the moment I set both AssetMovement attributes to None, as other exchanges do, and I'm not sure if we really want to request both endpoints and mix just for getting these 2 attributes.

You can search "crypto transactions" (below "user transactions") in their docs. Below an image.

![image](https://user-images.githubusercontent.com/14769962/101158390-8cdd4a00-3623-11eb-8abf-32dc180559e6.png)

Reply:
> and I'm not sure if we really want to request both endpoints and mix just for getting these 2 attributes.

Please make an issue about us doingthis at some point. Does not need to be part of this PR, but we can have an issue to remind us to improve the implementation.

Reply:
Issue is here: https://github.com/rotki/rotki/issues/1863

Resolving comment.